### ISSUE TITLE
Enhancement: Improve binary display in Number bases

### ISSUE THREAD

**Issue by MaCo117** [Qalculate/qalculate-qt#150]
Following suggestions relate to the "Number bases" widget which displays the current result in different bases.

1. Bit positions are usually indexed from zero. The hint positions should be 7-15-23-31 and so on. Check windows calculator, speedcrunch or similar for reference.
  
    <img src="https://github.com/Qalculate/qalculate-qt/assets/9208666/228a85a6-c08d-42df-ab5e-0d286930f6e4" width="400">

2. The digits in binary display (while flippable), are tiny and with no margins, making bit flipping cumbersome. Increasing font size helps somewhat, but that badly affects the rest of the app UI. Nobody does this better than speedcrunch - high contrast, immediately readable positions, consistent digit width. It's the only thing preventing me from switching at the moment.
    
    <img src="https://github.com/Qalculate/qalculate-qt/assets/9208666/5b02d4d2-528e-4774-8fdd-c38e801059ab" width="400">

Either way, love the tool - fantastic job!

**Comment by hakan-demirli** [comment]
I agree with all. I would like also to add that 2's complement display would be awesome.

![image](https://github.com/user-attachments/assets/206f4948-2a36-42a4-a4f4-eb7f5866eaf5)

-5 is not really helpful since there is no `-` in binary form. A more useful output would be:
`11111111111111111111111111111011`
or
`FFFFFFFB`

**Comment by hanna-kn** [comment]
> -5 is not really helpful since there is no `-` in binary form. A more useful output would be:
> `11111111111111111111111111111011`
> or
> `FFFFFFFB`

`11111111111111111111111111111011` is exactly what is displayed in your screenshot (?). Two's complement representation is by default enabled for binary numbers and can be enabled for hexadecimal in preferences (where two's complement for input can also be enabled).

**Comment by hakan-demirli** [comment]
My bad. I was looking for the hex version, not the binary one. All good now:
![image](https://github.com/user-attachments/assets/02d16c42-04f5-43a0-bfa2-d3717fc00fd5)

**Comment by g-v-egidy** [comment]
Please still consider improving the binary number display and editing. It is something that I need very often for microcontroller and network programming, where certain options are commonly encoded in certain bits that are defined in some specification or datasheet. So the datasheet says that bit 23 is responsible for feature X, you are seeing a certain 32 bit hex number in the debugger and want to quickly find out if bit 23 is enabled or not. Then manipulate it and output the full 32 bits of it again and so on.

From all the other features and usability, Qalculate beats kcalc hands down. But I really like the bit field display and bit editing in kcalc and this is why I usually reach for it instead of using Qalculate.

Here is a screenshot of kcalc 26.04.3:

<img width="729" height="336" alt="Image" src="https://github.com/user-attachments/assets/4c2a1feb-963d-4cd2-8457-a1091e741884" />

You can very clearly see the enabled bits in black and disabled bits in white. They are big enough to easily click them to toggle. They are grouped in blocks of 8 for better overview. Also note the small popup when you hover with the mouse over one bit, it shows the bit number (very important so that you don't have to count positions on the screen) and also the decimal weight of this bit, sometimes handy.

I would really like if you could add something similar to Qalculate, automatically enabled with the "programming keypad" or maybe a separate option.

**Comment by hanna-kn** [comment]
> Also note the small popup when you hover with the mouse over one bit, it shows the bit number (very important so that you don't have to count positions on the screen) and also the decimal weight of this bit, sometimes handy.

This has now been added in qalculate-qt.

**Comment by g-v-egidy** [comment]
Thank you very much for adding the popup.

Is the bigger clickable bitfield view also something you would consider adding and it is just a matter of development time? Or do you think adding this is out of scope, taking up too much space on the screen or similar?

**Comment by g-v-egidy** [comment]
Oh, one more point I just noticed:

When you have some result shown in the number bases section and click on one bit, the bit is toggled and this changes the data in the expression line to the new number with the changed bit. This is good and exactly what I want.

But the cursor is now at the left side of the new number in the expression line. So when I now want to use this number for some further calculation (like add something to it), I first have to move the cursor to the right before I can continue to type. So please move the cursor to the right after flipping bits.

Also maybe consider having the bitflip not override the entry in the history buffer, but add a separate history line to make the bitflip obvious when looking at the history. If there are multiple consecutive bitflips, probably combine them in one history line.

**Comment by hanna-kn** [comment]
> Is the bigger clickable bitfield view also something you would consider adding and it is just a matter of development time? Or do you think adding this is out of scope, taking up too much space on the screen or similar?

It is more a matter of coming up with an appropriate design. I would prefer not to add a separate view, but improve the usability of the existing view (possibly with a togglable mode). An initial way to improve the current design might simply be to use a bigger monospaced font

> But the cursor is now at the left side of the new number in the expression line. So when I now want to use this number for some further calculation (like add something to it), I first have to move the cursor to the right before I can continue to type. So please move the cursor to the right after flipping bits.

Fixed.

> Also maybe consider having the bitflip not override the entry in the history buffer, but add a separate history line to make the bitflip obvious when looking at the history. If there are multiple consecutive bitflips, probably combine them in one history line.

If you want to preserve to original expression, you should commit the result to history using `Enter` or `=`.

**Comment by g-v-egidy** [comment]
> It is more a matter of coming up with an appropriate design. I would prefer not to add a separate view, but improve the usability of the existing view (possibly with a togglable mode). An initial way to improve the current design might simply be to use a bigger monospaced font

Thanks for explaining. Yes, I think a bigger monospaced should go a long way and not require a complete code redesign.

 
> > But the cursor is now at the left side of the new number in the expression line. So when I now want to use this number for some further calculation (like add something to it), I first have to move the cursor to the right before I can continue to type. So please move the cursor to the right after flipping bits.
> 
> Fixed.

Thank you very much for fixing it this quick. I was already picking at the source to implement this myself as I thought this might be a good starting point into the code when you pushed your change.

Your fix does exactly what I mentioned and solves exactly the problem I had. But now that I more carefully observed the behavior of qalculate, I'm not sure if just positioning the cursor on the end, or positioning the cursor on the end plus also marking the whole expression window is the better and more consistent behavior: When pressing = you also mark the expression window, so that any further numbers pressed overwrite the expression, while some clever logic keeps the expression when pressing operators.

Changing a bit and showing it in the expression window has a bit of showing a result, so marking it would be consistent to also mark it. Also I'm a bit doubtful if there is a case where you want to add digits at the end of something you just bit-manipulated. I guess when entering something in this moment it is more likely that you want to do mass bit-edits and are entering something new so overwriting might be the better option.

While the cursor thing before was annoying, this is just fine-tuning and I'm myself not 100% sure which is the better way.

> > Also maybe consider having the bitflip not override the entry in the history buffer, but add a separate history line to make the bitflip obvious when looking at the history. If there are multiple consecutive bitflips, probably combine them in one history line.
> 
> If you want to preserve to original expression, you should commit the result to history using `Enter` or `=`.

You are right, I did not observe this properly enough. Sorry for bothering you.

**Comment by hanna-kn** [comment]
> But now that I more carefully observed the behavior of qalculate, I'm not sure if just positioning the cursor on the end, or positioning the cursor on the end plus also marking the whole expression window is the better and more consistent behavior: When pressing = you also mark the expression window, so that any further numbers pressed overwrite the expression, while some clever logic keeps the expression when pressing operators.

I've now modified the behaviour so that the selection is preserved if the whole expression was selected before the bit flip.

**Comment by hanna-kn** [comment]
> Yes, I think a bigger monospaced should go a long way and not require a complete code redesign.

I've come to the conclusion that use of a monospaced font would not necessary be an improvement (it depends on the specific font), especially considering that "tnum" is already used to ensure that all digits are of the same size. Instead I've increased both the letter spacing and font size by 10%, and use bold weight for set bits (if this does not change the size of the character, which might be the case with some fonts).

**Comment by g-v-egidy** [comment]
Thanks for continuing the work to improve this issue.

I just tested your newest iteration and it indeed improves it, but I couldn't get the ones shown as bold. I tried it with several different, but still reasonable, fonts. Set as "Custom application font". Then I patched your change to always set `use_bold_bin1=true;` and tried it again. Now it worked and looked definitely better, but the positioning slightly moved with each bit flip, due to the fonts having different sizes for bold.

Since this effect probably depends on the environment:
Fedora 43 x86_64
plasma-desktop-6.7.2-1.fc43.x86_64
qt6-qtbase-6.10.3-1.fc43.x86_64
xorg-x11-server-Xorg-21.1.24-1.fc43.x86_64

Fonts I tried: Noto Sans, Liberation Sans.

When I selected a dedicated monospace font for the whole application it worked without changing use_bold_bin1. But then the rest Qalculate wasn't pretty to use.

Maybe this would work better when you add a dedicated font option for the number bases display?

**Comment by hanna-kn** [comment]
> I just tested your newest iteration and it indeed improves it, but I couldn't get the ones shown as bold. I tried it with several different, but still reasonable, fonts. Set as "Custom application font". Then I patched your change to always set use_bold_bin1=true; and tried it again. Now it worked and looked definitely better, but the positioning slightly moved with each bit flip, due to the fonts having different sizes for bold.

This might be caused by full hinting. The application will now change the default hinting to vertical (light) hinting in an attempt to avoid this. I will (definitely) add an option to change the number bases font if this does not work.
